### PR TITLE
[IPv6] Check unspecified/loopback/multicast reserved/matching addresses in prvAllowIPPacketIPv6.

### PR DESCRIPTION
**Pull request by ActoryOu**
[IPv6] Check unspecified/loopback/multicast reserved/matching addresses in prvAllowIPPacketIPv6.

Description
-----------

From [RFC 4291](https://www.rfc-editor.org/rfc/rfc4291.html), some IPv6 addresses are reserved and shouldn't be used. Drop packets when the address met below conditions:
1. Source or destination address have unspecified address (All 0s).
2. Check endpoint's IP address with packet's destination.
3. Allow loopback packets from this node itself only.
4. Check if multicase address isn't valid. (Refer to [sec 2.7](https://www.rfc-editor.org/rfc/rfc4291.html#section-2.7) in RFC4291)
5. Fix coverity warnings.

Test Steps
-----------

Run protocol tests UDP/Others/003, UDP/Others/004, IPv6/datagram/010~017, 019.

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [ ] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: source/FreeRTOS_IPv6.c [hunk 1]
```diff
@@ -64,15 +130,34 @@ const struct xIPv6_Address FreeRTOS_in6addr_loopback = { { 0, 0, 0, 0, 0, 0, 0,
  */
 BaseType_t xIsIPv6Multicast( const IPv6_Address_t * pxIPAddress )
 {
-    BaseType_t xReturn;
+    BaseType_t xReturn = pdFALSE;
```
Comment (shubnil) [inline source/FreeRTOS_IPv6.c:133]: Instead of xIsIPv6Multicast, xIsIPv6AllowedMulticast seems more apt

File: source/FreeRTOS_IPv6.c [hunk 2]
```diff
@@ -55,6 +64,63 @@ const struct xIPv6_Address FreeRTOS_in6addr_any = { 0 };
...
+    }
+
+    return xReturn;
+}
+
+
+/*-----------------------------------------------------------*/
+
 /**
  * @brief Check whether this IPv6 address is a multicast address or not.
```
Comment (shubnil) [inline source/FreeRTOS_IPv6.c:125]: Please change the description as well to something like "Check whether this IPv6 address is an allowed multicast address or not." And add the description for return value as well.

File: source/FreeRTOS_IPv6.c [hunk 3]
```diff
@@ -55,6 +64,63 @@ const struct xIPv6_Address FreeRTOS_in6addr_any = { 0 };
...
+
+    configASSERT( pxIPv6Address != NULL );
+
+    ( void ) memcpy( &( xReturnGroupID.ucBytes[ 2 ] ), &( pxIPv6Address->ucBytes[ 2 ] ), 14 );
+
+    return xReturnGroupID;
+}
+
+/**
+ * @brief Check if the packet is a legal loopback packet.
```
Comment (shubnil) [inline source/FreeRTOS_IPv6.c:96]: I think, we do not need legal , we can just say "Check if the packet is a loopback packet." Same in line 100.

**Review by shubnil — CHANGES_REQUESTED** [review]

File: source/FreeRTOS_IPv6.c (hunk 1 above)
Comment (ActoryOu) [reply source/FreeRTOS_IPv6.c:133]: Agree. Let me rename it.

**Review by shubnil — APPROVED** [review]

**Review by moninom1 — APPROVED** [review]

**Pull request merged**